epilepsy12/common_view_functions/aggregate_by.py

```diff
@@ -345,80 +345,173 @@ def update_kpi_aggregation_model(
-                )
-                return
+            if open_access:
+                # for public view: create a new record
+
+                value_count["abstraction_relation"] = abstraction_relation_instance
+                value_count["cohort"] = cohort
+                value_count["open_access"] = open_access
+                
+                try:
+                    new_obj = AbstractionKPIAggregationModel.objects.create(**value_count)
+                    logger.debug(f"created {new_obj}")
+                except Exception as error:
+                    logger.exception(
+                        f"Can't save new KPIAggregations for {abstraction_level} for {abstraction_relation_instance}: {error}"
+                    )
+                    return
 
-            if created:
-                logger.debug(f"created {new_obj}")
             else:
-                logger.debug(f"updated {new_obj}")
+                
+                # not for public view - create or update existing
+                try:
+                    (
+                        new_obj,
+                        created,
+                    ) = AbstractionKPIAggregationModel.objects.update_or_create(
+                        defaults={
+                            "abstraction_relation": abstraction_relation_instance,
+                            "cohort": cohort,
+                            "open_access": False,
+                            **value_count,
+                        },
+                        abstraction_relation=abstraction_relation_instance,
+                        cohort=cohort,
+                        open_access=open_access,
+                    )
+                    logger.info(f"updating/saving: {abstraction_relation_instance}")
+                except Exception as error:
+                    logger.exception(
+                        f"CLOSED VIEW: Can't update/save KPIAggregations for {abstraction_level} for {abstraction_relation_instance}: {error}"
+                    )
+                    return
+
+                if created:
+                    logger.debug(f"created {new_obj}")
+                else:
+                    logger.debug(f"updated {new_obj}")
             
-    
-    logger.info(f"{len(list_of_updated_abstraction_level_instance)} scored {abstraction_level.name} instances updated with aggregated scores of a total {AbstractionKPIAggregationModel.objects.filter(cohort=cohort).count()} {abstraction_level.name}s")
+    # value_counts for this abstraction level  already updated
+    # Set all measures to 0 for remaining abstraction levels
+    logger.debug(f"{len(list_of_updated_abstraction_level_instance)} scored {abstraction_level.name} instances updated with aggregated scores of a total {AbstractionKPIAggregationModel.objects.filter(cohort=cohort).count()} {abstraction_level.name}s")
+    empty_kpis = {
```

Comment:
I think we can probably avoid writing the zero rows now we know that we must clear out all the aggregations rows and re-seed if we change the case filtering criteria at the top of the pipeline